charms/knative-serving/src/charm.py

```diff
@@ -33,9 +34,19 @@ def __init__(self, *args):
         self._app_name = self.app.name
         self._namespace = self.model.name
         self._resource_handler = None
+        # Instantiate the GatewayProvider class, one instance for sharing the local gateway
+        # another one for sharing the ingress gateway
+        self._ingress_gateway_provider = GatewayProvider(self, relation_name="ingress-gateway")
+        self._local_gateway_provider = GatewayProvider(self, relation_name="local-gateway")
 
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(
+            self.on["ingress-gateway"].relation_changed, self._on_ingress_gateway_relation_changed
+        )
+        self.framework.observe(
+            self.on["local-gateway"].relation_changed, self._on_local_gateway_relation_changed
+        )
```

Comment:
Should this be relation_created, not relation_changed?  Is there something we do here that reacts to information sent from the related applications?  

Functionally this does still work because we're guaranteed to get a relation_changed event after every relation_joined, but it feels odd to rely on that no?  We do this for other relations that use SDI, but that is because SDI does a version exchange which requires two-way communication.